### PR TITLE
Muxas/gqa

### PR DESCRIPTION
**Pull request by Muxas**
This PR implements LlamaAttention without RotaryEmbedding. It is checked against LlamaAttention from transformers by providing zeroed argument `position_ids`. @daskol please check if test can be further improved for readability and `pytest` ideology.

### CODE REVIEW

File: wrappers/python/tests/layer/test_llama_attention.py [hunk 1]
```diff
@@ -39,37 +37,77 @@ class LlamaAttentionTestParams:
...
+        n_emb=128,
+        n_emb_tile=32,
+        n_seq=64,
+        n_seq_tile=16,
+        n_batch=4,
+        n_batch_tile=1,
+        n_head=16,
+        n_head_tile=8,
+        n_head_kv=4,
+        dtype=np.dtypes.Float32DType(),
```
Comment (daskol) [inline wrappers/python/tests/layer/test_llama_attention.py:67]: Just `np.float32`.

File: wrappers/python/tests/layer/test_llama_attention.py [hunk 2]
```diff
@@ -79,48 +117,59 @@ def generate_inputs(params: LlamaAttentionTestParams):
...
     return torch_layer, nntile_layer, x_torch, y_grad_torch
 
+
 @pytest.mark.parametrize("params", TEST_PARAMS)
 class TestLlamaAttention:
-    def test_from_to_torch(self, starpu_simple, torch_rng, \
-            params: LlamaAttentionTestParams):
+    def test_from_to_torch(
+        self, starpu_simple, torch_rng, params: LlamaAttentionTestParams
+    ):
```
Comment (daskol) [inline wrappers/python/tests/layer/test_llama_attention.py:137]: I didn't get what `from_to_torch` mean. Does it test that layers are identically initialized, doesn't it?

**Review by daskol — COMMENTED** [review]
General comment is to sticker to `numpy.testing` rather than `torch.testing`. NumPy provides a richer set of assertions. Its core idea is a drop-in replacement for built-in `assert` statement with pretty printing and numerical precision controls, i.e.

```python
from numpy.testing import assert_allclose, assert_equal, ...
assert_equal(lhs, rhs)
```

Mixing different APIs could be potentially confusing for a reader. Also, inter-op in PyTorch is a quite lame.

File: wrappers/python/tests/layer/test_llama_attention.py (hunk 2 above)
Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py:137]: It means testing `from_torch` and `to_torch` functionality. Does `test_from_torch_and_to_torch` sound better?

File: wrappers/python/tests/layer/test_llama_attention.py (hunk 1 above)
Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: I just repeated how it is done in a batch norm test. We have to either use Torch dtypes, or rely on our own dtypes. Supporting bf16, fp16, fp8 and quantised formats is out of Numpy scope.

Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: #102 is to resolve this issue in the future

Comment (daskol) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: > Supporting bf16, fp16, fp8 and quantised formats is out of Numpy scope.

Absolutely not. There are widely-adopted custom types (e.g. `pandas`). Also, NumPy facilitates an extensions with [NEP-42][1]. Thanks to JAX team, we have [ml-dtypes][1] which provide in framework-agnostic way common low-bits floating types.

\[1\] [NEP-42][1]
\[2\] [numpy/numpy-user-dtypes][2]
\[3\] [jax-ml/ml-dtypes][3]

[1]: https://numpy.org/neps/nep-0042-new-dtypes.html
[2]: https://github.com/numpy/numpy-user-dtypes
[3]: https://github.com/jax-ml/ml_dtypes

File: wrappers/python/tests/layer/test_llama_attention.py (hunk 2 above)
Comment (daskol) [reply wrappers/python/tests/layer/test_llama_attention.py:137]: I would name it `test_torch_coercion` or `test_torch_interop` or `test_torch_casting`. Something like that.

Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py:137]: `test_torch_casting` sounds good

File: wrappers/python/tests/layer/test_llama_attention.py (hunk 1 above)
Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: NEP-42 is yet under construction, as the link says, but in the future it will allow us to add our own quantised formats. JAX ml-dtypes can be used for now, as it covers bfloat16 and float8 types. We need these only for testing purposes, as an interoperation between NNTile and Numpy is through upcasting NNTile data into a `float` for all 32 and less bit floats and into a `double` for 64-bit floats.

Comment (daskol) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: > We need these only for testing purposes, as an interoperation between NNTile and Numpy is through upcasting NNTile data into a float for all 32 and less bit floats and into a double for 64-bit floats.

Then we do not actually need a dtypes. Just define `upcast()` routine and expose it to Python with limited visiblity. That's all.

Comment (daskol) [reply wrappers/python/tests/layer/test_llama_attention.py:67]: This is a distinct topic. I don't understand why you brought it now.

Original comment was about replacing internal `numpy.dtypes.*` in favor of `numpy.float32`.

Comment (Muxas) [reply wrappers/python/tests/layer/test_llama_attention.py]: I just made it a string. Looks much simpler now

File: wrappers/python/nntile/layer/llama_attention.py [hunk 3]
```diff
@@ -0,0 +1,1330 @@
...
+        # Rotate axes into
+        # (head_size, n_seq, n_batch, kv_group_size, n_head_kv)
+        transpose_async(1.0, self.q_transposed.value, self.q.value, 2)
+        # X_Q, W_Q and Q_transposed can be offloaded from GPU
+        self.x_q.value.wont_use()
+        self.q_transposed.value.invalidate_submit()
+        self.w_q.value.wont_use()
+        # Apply bias if needed
+        if self.in_proj_bias_q is not None:
+            # batched add_fiber (head_size, batch=(kv_group_size, n_head_kv))
```
Comment (amkatrutsa) [inline wrappers/python/nntile/layer/llama_attention.py:656]: What does internal tuple batch=() mean? What shape should return in this case?

Comment (Muxas) [reply wrappers/python/nntile/layer/llama_attention.py:656]: This is just a virtual union of several dimensions into a single batch dimension. Parameter `batch_ndim=2` is set in the next line.
